airqo-platform/AirQo-api
# refactor data mgt

# Chore

**_WHAT DOES THIS PR DO?_**

- [x] This PR refactors the codebase for data management to accommodate all data engineering work (ETL functions in different languages). 

- [x] The same PR also introduces a new endpoint that shares more descriptive response parameters/keys for recent device measurements. Parameters like pm2_5, pm10, etc are used instead of field names. In case some value does not exist for a specific parameter, it is not returned in the response object.

- [x] At the same time, the same PR brings about changes in the way we handle GPS inaccuracies, uses caching mechanism instead of requiring user to first query a DB. 

- [x] Same PR modifies the implementation of caching - adds expiry policy, and handles all logic from code, no need to provide query parameters. 

- [x] Updates the app.yaml to a _staging_ name since these changes are meant for a new staging AppEngine link. 

**_WHAT JIRA STORY/TASK IS RELATED TO THIS PR?_**
[PLAT-263](https://airqoteam.atlassian.net/browse/PLAT-263?atlOrigin=eyJpIjoiNzM1MTNhNmUzNTAxNDI0N2JhNWY3OGUxYmE2MTA5ZGMiLCJwIjoiaiJ9)

**_WHAT IS THE LINK TO THE PR BRANCH_**
https://github.com/airqo-platform/AirQo-api/tree/chore-refactor-data-mgt

**_HOW DO I TEST OUT THIS PR?_**
`cd AirQo-api/src/data-mgt/node`
`npm install`
Ensure that you have Redis running on your machine -separate shell tab

> for Linux

`npm run stage-mac`

> for Windows

`npm run stage-pc`

**_WHICH ENDPOINTS SHOULD BE READY FOR TESTING?:_** 

- [x] Get Last Entry (descriptive response parameters):  http://localhost:3000/api/v1/data/feeds/transform/recent?channel={CHANNEL_ID}

- [x] Get Last Entry (non descriptive response parameters):  http://localhost:3000/api/v1/data/feeds/recent/{CHANNEL_ID}

- [x] Get last Channel entry age:  http://localhost:3000/api/v1/data/channels/age?channel={CHANNEL_ID}

- [x] Get last Sensor entry age:   http://localhost:3000/api/v1/data/channels/fields/age?sensor=pm2_5&channel={CHANNEL_ID}

- [x] Get channels:  http://localhost:3000/api/v1/data/channels

- [x] Get number of devices/channels:  http://localhost:3000/api/v1/data/channels/count 

- [x] Get hourly feeds:  http://localhost:3000/api/v1/data/hourly/feeds?channel={CHANNEL_ID}


[API documentation](https://docs.google.com/spreadsheets/d/1-NGTDBtTcIi8Fg2UYSWjcK6geKrZdopz5kxRFsCEMPs/edit#gid=757070984)

SCREENSHOTS:
![image](https://user-images.githubusercontent.com/1590213/97402657-f4f18f80-1903-11eb-9114-f33e187a6d08.png)

![image](https://user-images.githubusercontent.com/1590213/97402664-f8851680-1903-11eb-9ea3-ea934f71c2e8.png)

![image](https://user-images.githubusercontent.com/1590213/97402674-fcb13400-1903-11eb-9cf6-8b91db237ef8.png)

![image](https://user-images.githubusercontent.com/1590213/97402681-ff138e00-1903-11eb-8617-a1c15af664e7.png)

![image](https://user-images.githubusercontent.com/1590213/97402686-020e7e80-1904-11eb-8510-e53dcfa9ec96.png)

![image](https://user-images.githubusercontent.com/1590213/97402694-05a20580-1904-11eb-8e59-5bf3bc5f7d64.png)

![image](https://user-images.githubusercontent.com/1590213/97405358-91b62c00-1908-11eb-85bc-67d991c64670.png)

**_IS THERE ANY JENKINS CONSOLE LINK FOR CODE COVERAGE AND BUILD INFO?_**
N/A

**_ARE THERE ANY RELATED PRs?_**
N/A
